src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,625 @@
+    under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    systemd is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sys/ioctl.h>
+#include <net/if.h>
+
+#include "alloc-util.h"
+#include "parse-util.h"
+#include "fd-util.h"
+#include "strv.h"
+#include "hexdecoct.h"
+#include "string-util.h"
+#include "wireguard.h"
+#include "networkd-link.h"
+#include "networkd-util.h"
+#include "networkd-manager.h"
+
+#define WG_SET_DEVICE (SIOCDEVPRIVATE + 1)
+#define MAX_CONCURRENT_QUERIES 2048
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *wireguard_peer_new(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        if (!peer)
+                return NULL;
+        peer->fields.persistent_keepalive_interval = (uint16_t)-1;
+
+        LIST_PREPEND(peers, w->peers, peer);
+        w->allocation_size += sizeof(struct wgpeer);
+        w->last_peer_section = section;
+        w->dev.num_peers++;
+
+        return peer;
+}
+
+static int update_wireguard_config(NetDev *netdev) {
+        _cleanup_close_ int fd = -1;
+        _cleanup_free_ void *data = NULL;
+        uint8_t *pos = NULL;
+        struct ifreq ifreq = {};
+        Wireguard *w;
+        WireguardPeer *peer;
+        WireguardIpmask *mask;
+
+        assert(netdev);
+
+        w = WIREGUARD(netdev);
+
+        assert(w);
+
+        fd = socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0)
+                return log_netdev_error_errno(netdev, errno, "Failed to open AF_INET to configure wireguard: %m");
+
+        strncpy(ifreq.ifr_name, netdev->ifname, IFNAMSIZ);
+        data = new0(char, w->allocation_size);
+        if (!data)
+                return log_oom();
+
+        pos = data;
+        memcpy(pos, &w->dev, sizeof(struct wgdevice));
+        pos = pos + sizeof(struct wgdevice);
+        LIST_FOREACH(peers, peer, w->peers) {
+                memcpy(pos, &peer->fields, sizeof(struct wgpeer));
+                pos = pos + sizeof(struct wgpeer);
+
+                LIST_FOREACH(ipmasks, mask, peer->ipmasks) {
+                        memcpy(pos, &mask->fields, sizeof(struct wgipmask)),
+                        pos = pos + sizeof(struct wgipmask);
+                }
+        }
+        assert((size_t)((char*)pos - (char*)data) == w->allocation_size);
```

Comment:
So essentially lines 85-97 do a single memcpy of size `w->allocation_size`. Why not just do a single memcpy? If it is still wanted, the check that the sizes add up correctly can be put under an `#ifndef NDEBUG` stanza.

Reply:
This part will be also obsolete with netlink.